File: android/src/main/kotlin/ai/decart/vton/flutter/ChannelCodec.kt
package ai.decart.vton.flutter

import ai.decart.sdk.ConnectionState
import ai.decart.sdk.RealtimeModel
import ai.decart.sdk.realtime.ConnectionQuality
import ai.decart.sdk.realtime.ConnectivityTransport
import ai.decart.sdk.realtime.FacingMode
import ai.decart.sdk.realtime.MirrorMode
import ai.decart.sdk.realtime.RealtimeConfiguration
import ai.decart.sdk.realtime.Resolution

/**
 * The one place on the Android side that knows the platform-channel wire format.
 *
 * Its Dart counterpart is `lib/src/decart_vton_platform.dart` and its iOS
 * counterpart is `ChannelCodec.swift`. If you change a key here, change it in
 * both of those too — the contract is written out in IMPLEMENTATION.md.
 */
internal object ChannelCodec {

    // ── argument decoding ────────────────────────────────────────────────────

    fun requireArgs(arguments: Any?): Map<String, Any?> {
        @Suppress("UNCHECKED_CAST")
        return (arguments as? Map<String, Any?>)
            ?: throw VtonPluginException(
                ErrorCodes.INVALID_INPUT,
                "Expected a map of arguments but received ${arguments?.javaClass?.simpleName ?: "null"}.",
            )
    }

    fun Map<String, Any?>.string(key: String): String =
        this[key] as? String
            ?: throw VtonPluginException(
                ErrorCodes.INVALID_INPUT,
                "Missing required String argument '$key'.",
            )

    fun Map<String, Any?>.stringOrNull(key: String): String? = this[key] as? String

    fun Map<String, Any?>.int(key: String, fallback: Int): Int = when (val v = this[key]) {
        is Int -> v
        is Long -> v.toInt()
        is Number -> v.toInt()
        else -> fallback
    }

    fun Map<String, Any?>.long(key: String, fallback: Long): Long = when (val v = this[key]) {
        is Long -> v
        is Int -> v.toLong()
        is Number -> v.toLong()
        else -> fallback
    }

    fun Map<String, Any?>.bool(key: String, fallback: Boolean): Boolean =
        this[key] as? Boolean ?: fallback

    fun Map<String, Any?>.bytes(key: String): ByteArray? = this[key] as? ByteArray

    @Suppress("UNCHECKED_CAST")
    fun Map<String, Any?>.map(key: String): Map<String, Any?>? =
        this[key] as? Map<String, Any?>

    // ── enum mapping ─────────────────────────────────────────────────────────

    fun facing(wire: String?): FacingMode =
        if (wire == "back") FacingMode.BACK else FacingMode.FRONT

    fun facingToWire(facing: FacingMode): String =
        if (facing == FacingMode.BACK) "back" else "front"

    fun mirror(wire: String?): MirrorMode = when (wire) {
        "off" -> MirrorMode.OFF
        "on" -> MirrorMode.ON
        else -> MirrorMode.AUTO
    }

    fun resolution(wire: String?): Resolution? = when (wire) {
        "720p" -> Resolution.P720
        "1080p" -> Resolution.P1080
        else -> null
    }

    fun connectionStateToWire(state: ConnectionState): String = when (state) {
        ConnectionState.DISCONNECTED -> "disconnected"
        ConnectionState.CONNECTING -> "connecting"
        ConnectionState.CONNECTED -> "connected"
        ConnectionState.GENERATING -> "generating"
        ConnectionState.RECONNECTING -> "reconnecting"
    }

    /**
     * Maps the SDK's four-level quality scale onto the Dart five-level one.
     *
     * The extra Dart level is `unknown`, which the SDK expresses as "no report
     * yet" rather than as an enum case, so it is never produced here.
     * iOS's [ConnectionQuality] has the same four cases and maps identically.
     */
    fun qualityToWire(quality: ConnectionQuality): String = when (quality) {
        ConnectionQuality.GOOD -> "excellent"
        ConnectionQuality.FAIR -> "good"
        ConnectionQuality.POOR -> "poor"
        ConnectionQuality.CRITICAL -> "unusable"
    }

    fun transportToWire(transport: ConnectivityTransport): String =
        transport.name.lowercase()

    // ── configuration building ───────────────────────────────────────────────

    /**
     * Builds a [RealtimeModel] from the values Dart sent.
     *
     * Deliberately constructed from the wire payload rather than looked up in
     * `RealtimeModels`: the Dart [VtonModel] enum is the single source of truth
     * for dimensions and frame rate, and building the model here means a new
     * server-side model can be supported by editing one Dart enum instead of
     * two native registries.
     */
    fun realtimeModel(args: Map<String, Any?>): RealtimeModel = RealtimeModel(
        name = args.string("model"),
        urlPath = "/v1/stream",
        fps = args.int("fps", 30),
        width = args.int("width", 1088),
        height = args.int("height", 624),
    )

    fun realtimeConfiguration(args: Map<String, Any?>): RealtimeConfiguration {
        val video = args.map("video")
        return RealtimeConfiguration(
            connection = RealtimeConfiguration.ConnectionConfig(
                connectionTimeoutMs = args.long("connectTimeoutMs", 30_000L),
            ),
            media = RealtimeConfiguration.MediaConfig(
                video = RealtimeConfiguration.VideoConfig(
                    maxBitrate = video?.int("maxBitrate", 2_500_000) ?: 2_500_000,
                    maxFramerate = video?.int("maxFramerate", 30) ?: 30,
                    preferredCodec = video?.stringOrNull("preferredCodec") ?: "vp8",
                    simulcast = video?.bool("simulcast", true) ?: true,
                ),
            ),
        )
    }

    // ── event payloads ───────────────────────────────────────────────────────

    fun connectionStateEvent(state: ConnectionState): Map<String, Any?> =
        mapOf("type" to "connectionState", "state" to connectionStateToWire(state))

    fun sessionStartedEvent(sessionId: String, subscribeToken: String?): Map<String, Any?> =
        mapOf(
            "type" to "sessionStarted",
            "sessionId" to sessionId,
            "subscribeToken" to subscribeToken,
        )

    fun generationTickEvent(seconds: Double): Map<String, Any?> =
        mapOf("type" to "generationTick", "seconds" to seconds)

    fun remoteStreamEvent(): Map<String, Any?> = mapOf("type" to "remoteStreamUpdated")

    fun localStreamEvent(): Map<String, Any?> = mapOf("type" to "localStreamUpdated")

    fun errorEvent(code: String, message: String, details: Any? = null): Map<String, Any?> =
        mapOf(
            "type" to "error",
            "code" to code,
            "message" to message,
            "details" to details,
        )

    fun connectionQualityEvent(
        quality: ConnectionQuality,
        rttMs: Double?,
        packetLoss: Double?,
        jitterMs: Double?,
    ): Map<String, Any?> = mapOf(
        "type" to "connectionQuality",
        "quality" to qualityToWire(quality),
        "roundTripMs" to rttMs?.let { Math.round(it).toInt() },
        "packetLoss" to packetLoss,
        "jitterMs" to jitterMs?.let { Math.round(it).toInt() },
    )
}

File: ios/decart_vton_flutter/Sources/decart_vton_flutter/ChannelCodec.swift
import AVFoundation
import DecartSDK
import Flutter
import Foundation

/// The one place on the iOS side that knows the platform-channel wire format.
///
/// Its counterparts are `lib/src/decart_vton_platform.dart` and
/// `ChannelCodec.kt`. Change a key here and change it in both of those too —
/// the contract is written out in IMPLEMENTATION.md.
enum ChannelCodec {

    // MARK: - Argument decoding

    static func requireArgs(_ arguments: Any?) throws -> [String: Any] {
        if arguments == nil { return [:] }
        guard let map = arguments as? [String: Any] else {
            throw VtonPluginError(
                code: ErrorCodes.invalidInput,
                message: "Expected a map of arguments but received \(type(of: arguments))."
            )
        }
        return map
    }

    static func string(_ args: [String: Any], _ key: String) -> String? {
        guard let value = args[key] as? String, !value.isEmpty else { return nil }
        return value
    }

    static func int(_ args: [String: Any], _ key: String, _ fallback: Int) -> Int {
        if let value = args[key] as? Int { return value }
        if let value = args[key] as? NSNumber { return value.intValue }
        return fallback
    }

    static func bool(_ args: [String: Any], _ key: String, _ fallback: Bool) -> Bool {
        if let value = args[key] as? Bool { return value }
        if let value = args[key] as? NSNumber { return value.boolValue }
        return fallback
    }

    /// Standard-codec `Uint8List` arrives as `FlutterStandardTypedData`.
    static func bytes(_ args: [String: Any], _ key: String) -> Data? {
        if let typed = args[key] as? FlutterStandardTypedData {
            return typed.data.isEmpty ? nil : typed.data
        }
        if let data = args[key] as? Data {
            return data.isEmpty ? nil : data
        }
        return nil
    }

    static func map(_ args: [String: Any], _ key: String) -> [String: Any]? {
        args[key] as? [String: Any]
    }

    // MARK: - Enum mapping

    static func position(_ wire: String?) -> AVCaptureDevice.Position {
        wire == "back" ? .back : .front
    }

    static func mirror(_ wire: String?) -> MirrorMode {
        switch wire {
        case "off": return .off
        case "on": return .on
        default: return .auto
        }
    }

    static func resolution(_ wire: String?) -> Resolution? {
        switch wire {
        case "720p": return .p720
        case "1080p": return .p1080
        default: return nil
        }
    }

    /// Collapses the iOS SDK's seven-case state onto the shared wire vocabulary.
    ///
    /// All seven map one-to-one; the Dart enum was widened to the union of both
    /// platforms specifically so nothing has to be lost here.
    static func connectionStateToWire(_ state: DecartRealtimeConnectionState) -> String {
        switch state {
        case .idle: return "idle"
        case .connecting: return "connecting"
        case .connected: return "connected"
        case .generating: return "generating"
        case .reconnecting: return "reconnecting"
        case .disconnected: return "disconnected"
        case .error: return "error"
        }
    }

    /// Same four-to-five mapping as `ChannelCodec.qualityToWire` on Android.
    static func qualityToWire(_ quality: ConnectionQuality) -> String {
        switch quality {
        case .good: return "excellent"
        case .fair: return "good"
        case .poor: return "poor"
        case .critical: return "unusable"
        }
    }

    static func transportToWire(_ transport: ConnectivityTransport) -> String {
        transport.rawValue
    }

    // MARK: - Configuration building

    /// Builds a `ModelDefinition` from what Dart sent.
    ///
    /// Deliberately constructed rather than looked up in `Models.realtime(_:)`:
    /// the Dart `VtonModel` enum is the single source of truth for dimensions,
    /// frame rate and reference-image support, so a new server-side model needs
    /// one Dart edit rather than three.
    static func modelDefinition(_ args: [String: Any]) -> ModelDefinition {
        ModelDefinition(
            name: string(args, "model") ?? "lucy-vton-latest",
            urlPath: "/v1/stream",
            fps: int(args, "fps", 30),
            width: int(args, "width", 1088),
            height: int(args, "height", 624),
            hasReferenceImage: bool(args, "supportsReferenceImage", true)
        )
    }

    static func realtimeConfiguration(
        _ args: [String: Any],
        model: ModelDefinition,
        initialPrompt: DecartPrompt
    ) -> RealtimeConfiguration {
        let video = map(args, "video")
        return RealtimeConfiguration(
            model: model,
            initialPrompt: initialPrompt,
            resolution: resolution(string(args, "resolution")),
            connection: RealtimeConfiguration.ConnectionConfig(
                connectionTimeout: TimeInterval(int(args, "connectTimeoutMs", 30_000)) / 1000.0
            ),
            media: RealtimeConfiguration.MediaConfig(
                video: RealtimeConfiguration.VideoConfig(
                    maxBitrate: video.map { int($0, "maxBitrate", 2_500_000) } ?? 2_500_000,
                    maxFramerate: video.map { int($0, "maxFramerate", 30) } ?? 30,
                    // The iOS SDK defaults to h264 and Android to vp8. Dart
                    // pins vp8 on both so the same code behaves the same way;
                    // see VtonVideoConfig.
                    preferredCodec: video.flatMap { string($0, "preferredCodec") } ?? "vp8",
                    simulcast: video.map { bool($0, "simulcast", true) } ?? true
                )
            )
        )
    }

    // MARK: - Event payloads

    static func connectionStateEvent(_ state: DecartRealtimeConnectionState) -> [String: Any?] {
        ["type": "connectionState", "state": connectionStateToWire(state)]
    }

    static func sessionStartedEvent(sessionId: String) -> [String: Any?] {
        // The iOS SDK does not surface `subscribeToken` at v0.6.9 — Android
        // does. The key is still sent (as nil) so the payload shape matches.
        ["type": "sessionStarted", "sessionId": sessionId, "subscribeToken": nil]
    }

    static func generationTickEvent(seconds: Double) -> [String: Any?] {
        ["type": "generationTick", "seconds": seconds]
    }

    static func remoteStreamEvent() -> [String: Any?] {
        ["type": "remoteStreamUpdated"]
    }

    static func localStreamEvent() -> [String: Any?] {
        ["type": "localStreamUpdated"]
    }

    static func errorEvent(code: String, message: String) -> [String: Any?] {
        ["type": "error", "code": code, "message": message, "details": nil]
    }

    static func connectionQualityEvent(_ report: ConnectionQualityReport) -> [String: Any?] {
        [
            "type": "connectionQuality",
            "quality": qualityToWire(report.quality),
            "roundTripMs": report.metrics.rttMs.map { Int($0.rounded()) },
            "packetLoss": report.metrics.packetLoss,
            "jitterMs": report.metrics.upstreamJitterMs.map { Int($0.rounded()) }
        ]
    }
}
